src/unitxt/metrics.py

```diff
@@ -400,31 +699,180 @@ def process(self, stream: Stream, stream_name: Optional[str] = None) -> Generato
+                    f"This field is required for group based metric computation."
                 )
-                global_score.update(confidence_interval)
+            group_key = additional_inputs[self.grouping_field]
+            for field_name in field_names:
+                group_to_instance_scores[group_key][field_name].append(
+                    instance["score"]["instance"][field_name]
+                )
+
+        # a list where each element is a group (not an instance), and the dict corresponds to
+        # the aggregate scores of each field within that group
+        return [
+            {
+                "score": {
+                    "instance": {
+                        field_name: group_aggregation_func(scores[field_name])
+                        for field_name in field_names
+                    }
+                }
+            }
+            for scores in group_to_instance_scores.values()
+        ]
+
+    def get_group_aggregated_instance_scores_by_subgroups(
+        self,
+        instances: List[dict],
+        field_names: List[str],
+        group_aggregation_func,
+        subgroup_field: str,
+        baseline_name,
+    ):
+        """Return a list of group aggregation function value for group_mean_subgroup_comparison reduction.
+
+        Args:
+            instances: List of observation instances with instance-level scores (fields) computed.
+            field_names: List of instance score names in each instance to apply the aggregation function.
+            group_aggregation_func: Callable aggregation function with arguments baseline_scores and other_scores.
+            subgroup_field: name of field in instance additional_inputs that contains the subgroup identifier
+            baseline_name: value of subgroup_field that indicates a score belongs in baseline_scores.
+
+        Returns:
+            List of dicts, each corresponding to a group of instances (defined by grouping_field)
+        """
+        from collections import defaultdict
+
+        # two-level defaultdict: first is the grouping, second is the field name
+        # first list is instance scores for baseline group, second is for comparison group
+        group_to_instance_scores = defaultdict(lambda: defaultdict(lambda: [[], []]))
 
         for instance in instances:
-            yield instance
+            additional_inputs = instance["additional_inputs"]
+            for cc in [self.grouping_field, subgroup_field]:
+                if cc not in additional_inputs:
+                    raise ValueError(
+                        f"Missing '{cc}' from instance {instance}. "
+                        f"This field is required for group based metric computation."
+                    )
+            group_key = additional_inputs[self.grouping_field]
+            # indicator if is in the baseline group or not
+            is_baseline = str(additional_inputs[subgroup_field]) == baseline_name
```

Comment:
suggestion: maybe have `is_baseline` directly defined on the instance, next to the `group_id`? isn't this already in the data? that way you will only need to define the is_baseline field name. i think you even can just fix it to some value, maybe its name can be `is_group_baseline`?
I think its nice that these things can be customized, but it makes things a bit difficult to follow ..

Reply:
Similarly you could say that if one want to do grouping, it should simply set the `group_id` field, without having to specify the group id field name. Meaning, just fix `grouping_field = "group_id"`.. how does that sound?

Reply:
not a bad idea, I guess the user could make these changes in the dataset card, to create this column if it doesn't exist already.